docs/src/literate/ns_vs_diffeq.jl

```diff
@@ -0,0 +1,463 @@
+                    Kₑ[BlockIndex((v▄, p▄), (i, j))] += (divφ * ψ) * dΩ
+                    Kₑ[BlockIndex((p▄, v▄), (j, i))] += (ψ * divφ) * dΩ
+                end
+            end
+        end
+
+        # The last step in the element loop is to assemble `Kₑ` and `fe`
+        # into the global `K` and `f` with `assemble!`.
+        #+
+        assemble!(stiffness_assembler, celldofs(cell), Kₑ)
+    end
+    return K
+end;
+
+# ### Solution of the semi-discretized system via DifferentialEquations.jl
+# First we assemble the linear portions for efficiency. These matrices are
+# assumed to be constant over time.
+T = 10.0
+Δt₀ = 0.01
+Δt_save = 0.1
+
+M = create_sparsity_pattern(dh);
+M = assemble_mass_matrix(cellvalues_v, cellvalues_p, M, dh);
+
+K = create_sparsity_pattern(dh);
+K = assemble_stokes_matrix(cellvalues_v, cellvalues_p, ν, K, dh);
+
+# These are our initial conditions. We start from the zero solution, because it
+# is trivially admissible if the Dirichlet conditions are zero everywhere on the
+# Dirichlet boundary for t=0. Note that the time stepper is also doing fine if the
+# Dirichlet condition is non-zero and not too pathological.
+u₀ = zeros(ndofs(dh))
+apply!(u₀, ch);
+
+# At the time of writing this example we have no clean way to hook into the
+# nonlinear solver backend to apply the Dirichlet BCs. As a hotfix we override
+# the newton initialization. We cannot solve all emerging issues by developing
+# a customized newton algorithm here. This hack should only be seen as an
+# intermediate step towards integration with OrdinaryDiffEq.jl.
+function OrdinaryDiffEq.initialize!(nlsolver::OrdinaryDiffEq.NLSolver{<:NLNewton,true}, integrator)
+    ## This block is copy pasta from OrdinaryDiffEq
+    @unpack u,uprev,t,dt,opts = integrator
+    @unpack z,tmp,cache = nlsolver
+    @unpack weight = cache
+
+    cache.invγdt = inv(dt * nlsolver.γ)
+    cache.tstep = integrator.t + nlsolver.c * dt
+    OrdinaryDiffEq.calculate_residuals!(weight, fill!(weight, one(eltype(u))), uprev, u,
+                         opts.abstol, opts.reltol, opts.internalnorm, t);
+
+    # Before starting the nonlinear solve we have to set the time correctly.
+    # Note that ch is a global variable for now.
+    #+
+    update!(ch, cache.tstep);
+
+    # The update of u takes uprev + z or tmp + z most of the time, so we have
+    # to enforce Dirichlet BCs here. Note that these mutations may break the
+    # error estimators.
+    #+
+    apply!(uprev, ch)
+    apply!(tmp, ch)
+    apply_zero!(z, ch);
+
+    nothing
+end;
+
+# For the linear equations we can cleanly integrate with the linear solver
+# interface provided by the DifferentialEquations ecosystem. We use a direct
+# solver for simplicity, altough it comes with some issues. Implementing
+# GMRES with efficient preconditioner is left open for future work.
+mutable struct FerriteLinSolve{CH,F}
+    ch::CH
+    factorization::F
+    A
```

Comment:
```suggestion
    A::Union{Nothing, SparseMatrixCSC}
```
I guess, can also be parametrized, at least some information should be beneficial for the compiler, see https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations

Reply:
Also, for some reason DiffEq chokes if we do this.

Reply:
hm but then it should still be possible to constraint the type by some parametrization e.g. `Union{Nothing,T}` and T is then `<: Union{MatrixFree,AbstractMatrix}`. Having no constraint doesn't feel good to me, since beginner will take the code and most likely copy paste it

Reply:
No constraints should be fine here. I do not think that possible type instabilities here could lead to bottlenecks - usually either the evaluation of the right hand side, assembly or the solve is the bottleneck.

Reply:
It's primarly about providing good pratical code examples and having unconstrained types in structs is not advised according to the docs. I benchmarked it, its just 4s difference 56s vs 52s with parametrization. But still, 7% by a trivial change. 

In case you want to change it:
```julia
FerriteLinSolve(ch) = FerriteLinSolve(ch,lu,lu(sparse(ones(1,1))))

mutable struct FerriteLinSolve{CH,F,T<:Factorization}
    ch::CH
    factorization::F
    A::T
end
```

Reply:
Okay, I agree on this point, thanks!